frame/vesting/src/lib.rs

```diff
@@ -301,116 +362,383 @@ pub mod pallet {
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `schedule1_index`: index of the first schedule to merge.
+		/// - `schedule2_index`: index of the second schedule to merge.
+		#[pallet::weight(
+			T::WeightInfo::not_unlocking_merge_schedules(MaxLocksOf::<T>::get(), T::MaxVestingSchedules::get())
+			.max(T::WeightInfo::unlocking_merge_schedules(MaxLocksOf::<T>::get(), T::MaxVestingSchedules::get()))
+		)]
+		pub fn merge_schedules(
+			origin: OriginFor<T>,
+			schedule1_index: u32,
+			schedule2_index: u32,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			if schedule1_index == schedule2_index {
+				return Ok(());
+			};
+			let schedule1_index = schedule1_index as usize;
+			let schedule2_index = schedule2_index as usize;
+
+			let schedules = Self::vesting(&who).ok_or(Error::<T>::NotVesting)?;
+			let merge_action = VestingAction::Merge(schedule1_index, schedule2_index);
 
-			T::Currency::transfer(&source, &target, schedule.locked, ExistenceRequirement::AllowDeath)?;
+			let (schedules, locked_now) =
+				Self::exec_action(schedules.to_vec(), merge_action)?;
 
-			Self::add_vesting_schedule(&target, schedule.locked, schedule.per_block, schedule.starting_block)
-				.expect("user does not have an existing vesting schedule; q.e.d.");
+			Self::write_vesting(&who, schedules)?;
+			Self::write_lock(&who, locked_now);
 
 			Ok(())
 		}
 	}
 }
 
 impl<T: Config> Pallet<T> {
-	/// (Re)set or remove the pallet's currency lock on `who`'s account in accordance with their
-	/// current unvested amount.
-	fn update_lock(who: T::AccountId) -> DispatchResult {
-		let vesting = Self::vesting(&who).ok_or(Error::<T>::NotVesting)?;
+	// Create a new `VestingInfo`, based off of two other `VestingInfo`s.
+	// NOTE: We assume both schedules have had funds unlocked up through the current block.
+	fn merge_vesting_info(
+		now: T::BlockNumber,
+		schedule1: VestingInfo<BalanceOf<T>, T::BlockNumber>,
+		schedule2: VestingInfo<BalanceOf<T>, T::BlockNumber>,
+	) -> Result<Option<VestingInfo<BalanceOf<T>, T::BlockNumber>>, DispatchError> {
+		let schedule1_ending_block = schedule1.ending_block_as_balance::<T::BlockNumberToBalance, T>()?;
+		let schedule2_ending_block = schedule2.ending_block_as_balance::<T::BlockNumberToBalance, T>()?;
+		let now_as_balance = T::BlockNumberToBalance::convert(now);
+
+		// Check if one or both schedules have ended.
+		match (schedule1_ending_block <= now_as_balance, schedule2_ending_block <= now_as_balance) {
+			// If both schedules have ended, we don't merge and exit early.
+			(true, true) => return Ok(None),
+			// If one schedule has ended, we treat the one that has not ended as the new
+			// merged schedule.
+			(true, false) => return Ok(Some(schedule2)),
+			(false, true) => return Ok(Some(schedule1)),
+			// If neither schedule has ended don't exit early.
+			_ => {}
+		}
+
+		let locked = schedule1
+			.locked_at::<T::BlockNumberToBalance>(now)
```

Comment:
seems like you can consider passing `now_as_balance` directly to `locked_at` as well.

Reply:
It doesn't help because we compare `now` to `starting_block` which are both `BlockNumber`; we use the conversion for `BlockNumberToBalance::from(now - start_block)`.